_posts/2021-05-18-managing-react-application-state-with-context.md

```diff
@@ -0,0 +1,177 @@
+
+# Why you might not need Redux anymore
+
+When Facebook released React back in 2013, it also publicised the Flux pattern as a way of structuring your applications since it complimented the composable view components by utilising a unidirectional data flow to make systems predictable and easy to test.
+
+![The Flux Pattern diagram](/img/blog/react-state-management-context/flux-pattern.png)
+
+In the Flux Pattern, there is a centralised **Store** which contains state to be shown in the **View**. The only way to modify that state is through dispatching **Actions** through a **Reducer** function (not shown but is essentially the Dispatcher in the diagram) which updates the store by returning the new state.
+
+Redux was created as a way of formalising the Flux pattern within your React applications, it provided a structure for managing your _store_, _reducers_ and _actions_ and quickly became the defacto standard for managing state. This resulted in a healthy developer ecosystem being built around Redux including the popular Redux development tools which allowed developers to rewind through actions that have been applied to the store to aid in debugging. One issue Redux helped with is it allowed you to reduce **prop drilling**; where you would pass _component props_ through multiple levels of the component tree resulting in a somewhat negative developer experience.
+
+However, one of the main complaints about Redux has always been the amount of boilerplate code needed to get started. I've seen many teams reach for Redux as a default when starting a new application when the local state management tools built into React is enough for their use case; they will often spend more time setting up Redux than actually building their app. Another issue I've found with Redux is how it enforces a single store for all of your application state which leads to the store being convoluted and any attempts to organise this mess involve deeply nesting your state.
+
+By using Context, we can resolve some of these issues since we are not limited by the opinions enforced by Redux however it does still require some boilerplate code. The benefit of using something built into React is that we can choose to implement it at any time with little to no increase in our overall package size and it is something developers may be familiar with already.
+
+# Using Context for Application State
+
+Using the same principles that Redux was built on, we can manage our application state using something that is built into the React library, namely Context. In terms of developer experience, it is also beneficial for us to use TypeScript which allows us to set up our `State` and `Action` types and benefit from the type checking prior to runtime. In our example, we have a simple application that displays a *count* and allows us to *increment*, *decrement* and *set the value* manually. We begin by setting up each of these actions and a state variable to track the count.
```

Comment:
This is the area in the blog to really promote `Context`, people hear Context but I don't think people know what Context actually means.. Emphasis the meaning of Context... I.e. you can have a Context to manage your Apps Theme, Language, Table constructs, or as a Form manager etc. Otherwise people might just create 1 big Context or Contexts which have cross-contextual implementation (anti-pattern).